Add `mypy` to `pre-commit`

suhaibmujahid opened the pull request:
<!---
Please describe why and what this Pull Request is doing
-->


Resolves #1472 and resolves #1364






## Checklist

<!---
The following should be done (and marked as completed) when applicable. Please do not remove inapplicable items.
-->

- [ ] Type annotations added to new functions
- [ ] Docs added to functions touched in main classes
- [ ] Dry-run produced the expected results

marco-c commented:
There are a few pre-existing errors to fix and some library type stubs to install (adding them to the additional_dependencies).

We can avoid fixing the errors related to six if we simply get rid of it (#1364).

suhaibmujahid commented:
> There are a few pre-existing errors to fix and some library type stubs to install (adding them to the additional_dependencies).
>
> We can avoid fixing the errors related to six if we simply get rid of it (#1364).

I already dropped `six` and added `types-pytz` and `types-python-dateutil` to `additional_dependencies`. Now working to fix the rest of the errors.



Thank you for referring to issue #1364. I will link it to this PR so it will be closed when we land it.

marco-c reviewed (approved):
Thanks!
